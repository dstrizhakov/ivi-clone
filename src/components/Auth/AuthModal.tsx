import React, { FC, useEffect, useState } from 'react';
import styles from './AuthModal.module.scss';
import { CgClose } from 'react-icons/cg';
import { BiUser } from 'react-icons/bi';
import FullScreenModal from '@/components/Modals/FullScreenModal/FullScreenModal';
import { P } from '../P/P';
import BarGraph from '@/components/BarGraph/BarGraph';
import { useSession, signIn, signOut } from 'next-auth/react';
import { Button } from '../Button/Button';
import { BsPencil } from 'react-icons/bs';
import { AiOutlineEyeInvisible, AiOutlineEye } from 'react-icons/ai';
import { TbReload } from 'react-icons/tb';
import { SlSocialVkontakte, SlSocialGoogle } from 'react-icons/sl';
import { selectModal, setShowAuth } from '@/store/reducers/modals.slice';
import { useAppDispatch, useAppSelector } from '@/hooks/redux';
import { useTranslation } from 'react-i18next';
import { BtnA } from '../Button/Button.props';
import Link from 'next/link';
import { useGoogleLoginQuery, useLoginMutation } from '@/services/auth.api';
import { useRouter } from 'next/router';
import { logout, selectAuth, setUser } from '@/store/reducers/auth.slice';
import { REGEX_EMAIL, REGEX_PASSWORD } from '@/constants/Constants';

const AuthModal: FC = (): JSX.Element => {
  const { t } = useTranslation();

  const router = useRouter();
  const [progress, setProgress] = useState<number>(5);
  const [step, setStep] = useState<number>(1);
  const [login, setLogin] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [showPassword, setShowPassword] = useState(false);
  const { showAuth } = useAppSelector(selectModal);
  const dispatch = useAppDispatch();
  const close = () => {
    dispatch(setShowAuth(false));
    setStep(() => 1);
  };
  const { user } = useAppSelector(selectAuth);

  const { data: session } = useSession();
  const [loginFunc] = useLoginMutation();
  const { data: googleLogin } = useGoogleLoginQuery();

  const logoutFunc = () => {
    signOut().then(() => {
      dispatch(logout());
      router.push('/profile').then(() => {});
    });
  };

  const nextStep = () => {
    if (step < 4) {
      setStep((prev) => prev + 1);
    }
  };

  const previousStep = () => {
    if (step >= 1) {
      setStep((prev) => prev - 1);
    }
  };
  const toggleShowPassword = () => {
    setShowPassword((prevState) => !prevState);
  };

  useEffect(() => {
    switch (step) {
      case 1:
        setProgress(5);
        break;
      case 2:
        setProgress(50);
        break;
      case 3:
        setProgress(75);

        loginFunc({ email: login, password })
          .unwrap()
          .then((res) => {
            setProgress(100);
            dispatch(setUser(res));
            close();
            setPassword(() => '');
            setLogin(() => '');
          })
          .catch((rejected) => console.error(rejected));
        break;
    }
  }, [step]);

  async function handleGoogleSingIn() {
    await signIn('google', { callbackUrl: `${process.env.NEXT_PUBLIC_URL}/profile` });
  }
  async function handleVkSingIn() {
    await signIn('vk', { callbackUrl: `${process.env.NEXT_PUBLIC_URL}/profile` });
  }

  async function handleAuth() {
    const credentials = { email: login, password };
    console.log('handleAuth', login, password);
    signIn('credentials', {
      ...credentials,
      // redirect: false,
      callbackUrl: `${process.env.NEXT_PUBLIC_URL}/profile`,
    })
      .then((response) => {
        console.log(response);
      })
      .catch((error) => {
        console.log(error);
      });
  }

  return (
    <FullScreenModal isOpen={showAuth} closeModal={close}>
      <div className={styles.chat}>
        <div className={styles.chat__header}>
          {step > 1 ? (
            <div className={styles.chat__welcome}>
              <h5 className={styles.chat__title}>{t('sections.hello')}</h5>
              {login && <P size="S">{login}</P>}
            </div>
          ) : (
            <div className={styles.chat__welcome}>
              <h5 className={styles.chat__title}>{t('buttons.login-signup')}</h5>
            </div>
          )}
          <div className={styles.chat__close}>
            <CgClose onClick={close} />
          </div>
          <div className={styles.chat__progress}>
            <BarGraph width={progress} />
          </div>
        </div>
        <div className={styles.chat__body}>
          {user ? (
            <div className={styles.chat__message}>
              <h1 onClick={() => logoutFunc()} title={'Нажмите, чтобы выйти из аккаунта'}>
                {t('sections.already-signed')}
              </h1>
            </div>
          ) : (
            <>
              {step >= 1 && (
                <div className={styles.chat__message}>
                  <h5>{t('buttons.login-signup-person')}</h5>
                  {step <= 1 && <P>{t('sections.use-service-any-device')}</P>}
                </div>
              )}
              {step < 2 ? (
                <div className={styles.input}>
                  <BiUser className={styles.input__icon} />
                  <input
                    type="text"
                    value={login}
                    onChange={(e) => setLogin(e.target.value)}
                    className={!!login ? styles.input__active : ''}
                  />
                  <label>{t('buttons.email-or-phone')}</label>
                </div>
              ) : (
                <div className={styles.chat__row}>
                  <Button appearance={BtnA.circle} onClick={() => previousStep()}>
                    <BsPencil />
                  </Button>
                  <div className={styles.chat__answer}>
                    <h5>{login}</h5>
                  </div>
                </div>
              )}
              {step >= 2 && (
                <>
                  <div className={styles.chat__message}>
                    <h5>
                      {t('buttons.enter-password')} {t('buttons.to-login')}
                    </h5>
                  </div>
                  <div className={`${styles.input} ${styles.password}`}>
                    <input
                      type={showPassword ? 'text' : 'password'}
                      value={password}
                      onChange={(e) => setPassword(e.target.value)}
                      className={!!password ? styles.input__active : ''}
                    />
                    <label>{t('buttons.enter-password')}</label>
                    {!showPassword ? (
                      <AiOutlineEye
                        className={`${styles.input__show} ${
                          //todo: менять цвет при невалидных данных (!login.match(REGEX_EMAIL))
                          !!password && styles.input__showActive
                        }`}
                        onClick={toggleShowPassword}
                      />
                    ) : (
                      <AiOutlineEyeInvisible
                        className={`${styles.input__show} ${
                          !!password && styles.input__showActive
                        }`}
                        onClick={toggleShowPassword}
                      />
                    )}
                  </div>
                </>
              )}
              {step < 2 ? (
                <button
                  disabled={!login.match(REGEX_EMAIL)}
                  className={styles.button}
                  onClick={nextStep}
                >
                  {t('buttons.continue')}
                </button>
              ) : (
                <button
                  disabled={!password.match(REGEX_PASSWORD)}
                  className={styles.button}
                  onClick={nextStep}
                >
                  {t('buttons.login')}
                </button>
              )}
              {step < 2 ? (
                <>
                  <div className={styles.chat__oauth}>
                    <Link
                      href={'http://localhost:3001/auth/google/login'}
                      target={'_blank'}
                      className={styles.button}
                    >
                      <span>{t('buttons.login-with')} Google backend</span>
                      <SlSocialGoogle />
                    </Link>
                    <button className={styles.button} onClick={() => handleGoogleSingIn()}>
                      <span>{t('buttons.login-with')} Google nextauth</span>
                      <SlSocialGoogle />
                    </button>
                    <button className={styles.button} onClick={() => handleVkSingIn()}>
                      <span>{t('buttons.login-with')} VK</span>
                      <SlSocialVkontakte />
                    </button>
                  </div>
                  <div className={styles.chat__confidential}>
                    <p>{t('sections.click-continue-agree')}</p>
                    <p>
                      <span>{t('descriptions.with')} </span>
                      <Link
                        href="https://www.ivi.tv/info/confidential"
                        target="_blank"
                        rel="noreferrer"
                      >
                        {t('sections.privacy-policy')}
                      </Link>
                    </p>
                    <p>
                      <span>{t('sections.and')} </span>
                      <Link
                        href="https://www.ivi.tv/info/agreement"
                        target="_blank"
                        rel="noreferrer"
                      >
                        {t('sections.user-agreement')}
                      </Link>
                    </p>
                  </div>
                </>
              ) : (
                <div className={styles.chat__recover}>
                  <TbReload />
                  <h5>{t('buttons.reset-password')}</h5>
                </div>
              )}
            </>
          )}
        </div>
      </div>
    </FullScreenModal>
  );
};

export default AuthModal;
