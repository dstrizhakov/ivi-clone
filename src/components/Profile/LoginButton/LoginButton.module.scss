.login {
  display: flex;
  button {
    padding: 20px 15px;
  }
}
