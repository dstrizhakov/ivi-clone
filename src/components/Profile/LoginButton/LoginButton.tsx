import React from 'react';
import styles from './LoginButton.module.scss';
import { Button } from '@/components/Button/Button';
import { BtnA, BtnS } from '@/components/Button/Button.props';
import { TiUserOutline } from 'react-icons/ti';
import { useAppDispatch } from '@/hooks/redux';
import { setShowAuth } from '@/store/reducers/modals.slice';
import { useTranslation } from 'react-i18next';

const LoginButton = () => {
  const { t } = useTranslation();

  const dispatch = useAppDispatch();
  const openLoginModal = () => {
    dispatch(setShowAuth(true));
  };
  return (
    <div className={styles.login} data-testid={'login-button'}>
      <Button
        onClick={() => openLoginModal()}
        size={BtnS.S}
        appearance={BtnA.red}
        title={t('buttons.login-signup') || 'Войти или зарегистрироваться'}
      >
        <TiUserOutline />
        {t('buttons.login-signup')}
      </Button>
    </div>
  );
};

export default LoginButton;
