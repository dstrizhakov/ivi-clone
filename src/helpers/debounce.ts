export const debounce = <Fn extends (...args: unknown[]) => unknown>(fn: Fn, ms: number) => {
  let timeout: NodeJS.Timeout;

  const debounced = (...args: unknown[]) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => fn(...args), ms);
  };

  return debounced as (...args) => ReturnType<Fn>;
};
