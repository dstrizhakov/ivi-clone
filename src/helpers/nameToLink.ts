export const nameToLink = (str: string): string => {
  return str.toLowerCase().split(' ').join('-');
};

export const reverseNameToLink = (str: string): string => {
  return str
    .split('-')
    .map((i) => i[0].toUpperCase() + i.slice(1))
    .join(' ');
};
