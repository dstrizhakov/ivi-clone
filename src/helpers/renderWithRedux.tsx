import { Provider } from 'react-redux';
import { store } from '@/store/store';
import { render } from '@testing-library/react';

export const renderWithRedux = (component) => {
  return render(<Provider store={store}>{component}</Provider>);
};
