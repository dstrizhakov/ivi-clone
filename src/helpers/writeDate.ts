export const writeDate = (date: number): string => {//DD.MM.YYYY
  const now: Date = new Date(date);
  const day = now.getDate();
  const month = now.getMonth() + 1;
  const year = now.getFullYear();
  return `${day < 10 ? `0${day}` : day}.${month < 10 ? `0${month}` : month}.${year}`;
};
