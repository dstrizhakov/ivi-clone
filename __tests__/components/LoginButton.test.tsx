import { renderWithRedux } from '../../src/helpers/renderWithRedux';
import LoginButton from '../../src/components/Profile/LoginButton/LoginButton';
import { fireEvent, screen } from '@testing-library/react';
import AuthModal from '../../src/components/Auth/AuthModal';

describe('login button', () => {
  //todo: fix
  test('login', async () => {
    renderWithRedux(
      <>
        <LoginButton />
        <AuthModal />
      </>
    );
    const button = screen.getByTestId('login-button');
    fireEvent.click(await button);
    await expect(screen.getByText(/войти/gi)).not.toBeNull();
  });
});
