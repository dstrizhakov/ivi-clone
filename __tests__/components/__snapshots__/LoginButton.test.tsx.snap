// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login button login 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login"
      >
        <button
          class="button red S"
          title="buttons.login-signup"
        >
          <svg
            baseProfile="tiny"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            version="1.2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 6c1.654 0 3 1.346 3 3s-1.346 3-3 3-3-1.346-3-3 1.346-3 3-3m0-2c-2.764 0-5 2.238-5 5s2.236 5 5 5 5-2.238 5-5-2.236-5-5-5zM12 17c2.021 0 3.301.771 3.783 1.445-.683.26-1.969.555-3.783.555-1.984 0-3.206-.305-3.818-.542.459-.715 1.777-1.458 3.818-1.458m0-2c-3.75 0-6 2-6 4 0 1 2.25 2 6 2 3.518 0 6-1 6-2 0-2-2.354-4-6-4z"
            />
          </svg>
          buttons.login-signup
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login"
    >
      <button
        class="button red S"
        title="buttons.login-signup"
      >
        <svg
          baseProfile="tiny"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          version="1.2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 6c1.654 0 3 1.346 3 3s-1.346 3-3 3-3-1.346-3-3 1.346-3 3-3m0-2c-2.764 0-5 2.238-5 5s2.236 5 5 5 5-2.238 5-5-2.236-5-5-5zM12 17c2.021 0 3.301.771 3.783 1.445-.683.26-1.969.555-3.783.555-1.984 0-3.206-.305-3.818-.542.459-.715 1.777-1.458 3.818-1.458m0-2c-3.75 0-6 2-6 4 0 1 2.25 2 6 2 3.518 0 6-1 6-2 0-2-2.354-4-6-4z"
          />
        </svg>
        buttons.login-signup
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
